File: StaticSharp/Modifiers/BackgroundImage/AbstractBackground.cs
using StaticSharp.Gears;

namespace StaticSharp {



    public enum BackgroundImageRepeat {
        Repeat,     // The background image is repeated both vertically and horizontally
        RepeatX,    // The background image is repeated only horizontally
        RepeatY,    // The background image is repeated only vertically
        NoRepeat,   // The background-image is not repeated. The image will only be shown once
        Space,      // The background-image is repeated as much as possible without clipping. The first and last image is pinned to either side of the element, and whitespace is distributed evenly between the images
        Round       // The background-image is repeated and squished or stretched to fill the space (no gaps)
    }

    public enum BlendMode {
        Normal,
        Multiply,
        Screen,
        Overlay,
        Darken,
        Lighten,
        ColorDodge,
        ColorBurn,
        HardLight,
        SoftLight,
        Difference,
        Exclusion,
        Hue,
        Saturation,
        Color,
        Luminosity
    }


    public interface JAbstractBackground : JModifier {
        bool Enabled { get; set; }
        string RawImage { get; set; }
        double X { get; set; }
        double Y { get; set; }
        double Width { get; set; }
        double Height { get; set; }

        
        BackgroundImageRepeat Repeat { get; set; }
        BlendMode BlendMode { get; set; }
    }


    [ConstructorJs]
    public partial class AbstractBackground : Modifier {
    }

}

File: StaticSharp/Modifiers/BackgroundImage/AbstractBackground.js
StaticSharpClass("StaticSharp.AbstractBackground", (modifier, element) => {
    StaticSharp.Modifier(modifier, element)

    


    modifier.Reactive = {
        Enabled: true,
        RawImage: undefined,

        X: 0,
        Y: 0,
        Width: undefined,
        Height: undefined,

        Repeat: "repeat",
        BlendMode: "normal"
    }


    modifier.getBackground = function () {
        let width = modifier.Width
        let widthCss = (width === undefined) ? "auto" : width + "px"
        let height = modifier.Height
        let heightCss = (height === undefined) ? "auto" : height + "px"

        return {
            backgroundImage: modifier.RawImage,
            backgroundPosition: `${modifier.X}px ${modifier.Y}px`,
            backgroundSize: widthCss + " " + heightCss,
            backgroundRepeat: CamelToKebab(modifier.Repeat),
            backgroundBlendMode: CamelToKebab(modifier.BlendMode)
        }
    }
    

    new Reaction(() => {
        var coModifiers = element.Modifiers.filter(x => x.is("StaticSharp.AbstractBackground") && x.Enabled)
        if (coModifiers.length == 0)
            return
        if (coModifiers[0] == modifier) {
            let data = coModifiers.map(x => x.getBackground()).reverse()
            for (let i of Object.keys(data[0]))
                element.style[i] = data.map(x => x[i]).join(',')
        }
    })

})